BidsInputs Refactoring

pvandyken opened the pull request:
## Proposed changes

Split `BidsInputs` into two classes: a smaller class describing a single dataset component and containing the lists, zip_lists, wildcard dict, and path, and a wrapper class compiling them together that takes the same API as `BidsInputs`. The inner class I've called `BidsComponent` and the wrapper I've called `BidsDataset`. The alternative name for the wrapper was `BidsComponents` (pluralized), but I found it confusing from a documentation perspective, and I think `BidsDataset` is more descriptive.

This new structure cleans up the implementation details in a few ways:

1. The old internal `_BidsLists` named tuple is no longer necessary and has been removed.
2. `input_lists` and `input_wildcards` are now defined in terms of `input_zip_lists`. This will make modification of the dataset (e.g. filtering) much easier, as now only one entry needs to be modified.
3. `BidsComponent` is equipped with an `__eq__()` operator, making testing much easier. Most of the tests in `test_generate_inputs.py` have been modified to take advantage of this new behaviour

Finally, this opens up new ways of interacting with datasets. The previous syntax (e.g. `inputs.input_lists["t1w"]`) will still work, but because `BidsDataset` is implented as a dict, a reverse syntax is also possible (e.g. `inputs["t1w"].input_lists`). This new syntax is possibly more intuitive.

## Backward Compatibility

As `BidsInputs` was always the return of `generate_inputs()` and never intended to be directly accessed, this change should not affect any existing workflows. If anyone feels it necessary, it should be possible to have a stub `BidsInputs` whose `__new__()` returns a copy of `BidsDataset`.

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you are unsure about any of the choices, don't hesitate to ask!

- [x] Changes have been tested to ensure that fix is effective or that a feature works.
- [x] Changes pass the unit tests
- [x] I have included necessary documentation or comments (as necessary)
- [x] Any dependent changes have been merged and published

tkkuehn commented on snakebids/core/input_generation.py:
```diff
@@ -75,19 +46,19 @@ class BidsInputsDict(TypedDict):
-    """Bids input entities and their resolved values
+@attr.define
+class BidsComponent:
+    """Component of a BidsDataset mapping entities to their resolved values
```
It would probably be good to note in the docstring that this is immutable.

pvandyken replied:
Actually, if we wanted full immutability, we could use this [immutabledict](https://github.com/corenting/immutabledict). Do people think it's worth it?

pvandyken replied:
I'm going to branch this into its own issue

tkkuehn commented on snakebids/tests/test_generate_inputs.py:
```diff
@@ -29,6 +31,188 @@
+
+@st.composite
+def bids_input(draw: st.DrawFn):
+    alphabet = st.characters(whitelist_categories=["Ll", "Lu", "Nd"])
```
I think you can use `BIDS_ALPHABET` here.

tkkuehn commented on snakebids/core/input_generation.py:
```diff
@@ -96,12 +67,119 @@ class BidsInputs:
-    input_wildcards: Dict[str, Dict[str, str]]
+    input_name: str = attr.field(on_setattr=attr.setters.frozen)
+    input_path: str = attr.field(on_setattr=attr.setters.frozen)
+    input_zip_lists: Dict[str, List[str]] = attr.field(on_setattr=attr.setters.frozen)
```
I think it would be worth adding a validator to make sure that all the zip lists are the same length. Also a validator to make sure the zip lists and input path have the same wildcards.

tkkuehn commented on snakebids/core/input_generation.py:
```diff
@@ -143,60 +223,38 @@ def subj_wildcards(self):
         -------
-        Tuple of Dicts
-            Tuple of all the dicts generated from the BidsLists
+        BidsDatasetDic
```
Typo

tkkuehn reviewed (changes requested):
Overall looks good, just a couple of suggestions.

pvandyken commented on snakebids/core/input_generation.py:
```diff
@@ -75,19 +46,19 @@ class BidsInputsDict(TypedDict):
+class BidsComponent:
+    """Component of a BidsDataset mapping entities to their resolved values
 
     Attributes
```
Hmm good catch... Actually, I think technically `input_lists` and `input_wildcards` should be documented under their respective properties (according to the sphinx docstring standards). I'm never fully sure how to document the attributes/parameters of an attrs class, because they're usually the same thing. My opinion for this is that since this class is not meant to be directly constructed, we should list them as attributes rather than parameters, but it's open to debate

pvandyken replied:
Hmm good catch... Actually, I think technically `input_lists` and `input_wildcards` should be documented under their respective properties (according to the sphinx docstring standards). I'm never fully sure how to document the attributes/parameters of an attrs class, because they're usually the same thing. My opinion for this is that since this class is not meant to be directly constructed, we should list them as attributes rather than parameters, but it's open to debate

pvandyken replied:
Cool, thanks for the link

pvandyken commented:
@tkkuehn I've addressed the comments you left. In addition, I've refactored and organized the testing code, moving all my custom hypothesis strategies to a new module.

tkkuehn commented on snakebids/core/input_generation.py:
```diff
@@ -75,33 +48,172 @@ class BidsInputsDict(TypedDict):
+
+    @input_zip_lists.validator  # type: ignore
+    def __(self, _, value: Dict[str, List[str]]):
+        lengths = [len(val) for val in value.values()]
```
This would be a bit simpler as a set comprehension.

tkkuehn commented on snakebids/core/input_generation.py:
```diff
@@ -75,33 +48,172 @@ class BidsInputsDict(TypedDict):
+    input_zip_lists: Dict[str, List[str]] = attr.field(on_setattr=attr.setters.frozen)
+
+    @input_zip_lists.validator  # type: ignore
+    def __(self, _, value: Dict[str, List[str]]):
```
My slight preference would be for this to have an informative name (still prepended with an underscore or two)

tkkuehn reviewed (approved):
A couple of nitpicks that you can address if you agree but I'm otherwise happy with this and don't need to see it again. Thanks again for your work on this, it's really nice!

pvandyken commented:
@akhanf, @kaitj, did you still want to take a look, or should I merge?

kaitj commented:
> 



> @akhanf, @kaitj, did you still want to take a look, or should I merge?

Just took another look at it - looks good to me.
